networkx/algorithms/dag.py

```diff
@@ -355,6 +387,108 @@ def transitive_closure(G):
 
 
 @not_implemented_for('undirected')
+def transitive_reduction(G):
+    r"""
+    Returns transitive closure of a directed graph.
+
+    The transitive closure minimizes the number of edges in a graph while
+    maintaining reachability.  Formally, the transitive closure of G = (V,E) is
+    a graph G+ = (V,E+) such that for all directed paths from u to v in G there
+    is a directed path from u to v in G+ and if there is no graph with fewer
+    edges than G+ that satisfies the first condition. When the graph is acyclic
+    this is also the unique minimal graph. The transitive reduction of a cyclic
+    graph may not be unique.
+
+    Parameters
+    ----------
+    G : NetworkX DiGraph
+        Graph
+
+    Returns
+    -------
+    TR : NetworkX DiGraph
+        Graph
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is not directed
+
+    Examples
+    ---------
+
+    >>> G = nx.DiGraph([(0, 1), (0, 2), (0, 4),
+    ...                 (0, 3), (1, 3), (2, 4),
+    ...                 (3, 4), (2, 4), (2, 3)])
+    >>> TR = nx.algorithms.dag.transitive_reduction(G)
+    >>> list(TR.edges())
+    [(0, 1), (0, 2), (1, 3), (2, 3), (3, 4)]
+
+    References
+    ----------
+    .. [1] http://dept-info.labri.fr/~thibault/tmp/0201008.pdf
+    .. [2] https://en.wikipedia.org/wiki/Transitive_reduction#Computing_the_reduction_using_the_closure
+    """
+    def invert_map(mapping):
+        """ point each value to a list of its corresponding keys """
+        from collections import defaultdict
+        pair_list = [(val, key) for key, val in mapping.items()]
+        # Initialize dict of lists
+        groupid2_items = defaultdict(list)
+        # Insert each item into the correct group
+        for groupid, item in pair_list:
+            groupid2_items[groupid].append(item)
+        return groupid2_items
+
+    has_cycles = not nx.is_directed_acyclic_graph(G)
+    if has_cycles:
+        # To handle cycles convert to graph of SCCs, find transitive reduction
+        # and then make a directed cycle for every SCC.
+        G_scc = nx.condensation(G)
+    else:
+        G_scc = G
+
+    # Adapted from
+    # http://stackoverflow.com/questions/17078696/transitive-reduction-python
```

Comment:
StackOverflow content is distributed under a Creative Commons Attribution-ShareAlike 3.0 license.